lib/cookies/util.js

```diff
@@ -3,22 +3,28 @@
 const assert = require('node:assert')
 const { kHeadersList } = require('../core/symbols')
 
+/**
+ * @param {string} value
+ * @returns {boolean}
+ */
 function isCTLExcludingHtab (value) {
-  if (value.length === 0) {
+  const valueLength = value.length
+  if (valueLength === 0) {
     return false
   }
 
-  for (const char of value) {
-    const code = char.charCodeAt(0)
+  for (let i = 0; i < valueLength; ++i) {
+    const code = value.charCodeAt(i)
 
     if (
-      (code >= 0x00 || code <= 0x08) ||
-      (code >= 0x0A || code <= 0x1F) ||
+      (code >= 0x00 && code <= 0x08) ||
+      (code >= 0x0A && code <= 0x1F) ||
```

Comment:
This could be simplified to

```suggestion
      (code < 0x20 && code !== 0x09) ||
```

But i didnt see any perf gains. Any ideas @tsctx

Reply:
I would like you to try this approach.

```js

const invalidChar = 1 << "a".charCodeAt(0) | 1 << "b".charCodeAt(0) // invalidChar chars.


// usage
const c = char.charCodeAt(0)
if (
  // check c is ascii code.
  (c & 0x7F) === c &&
  // check c is not a invalidChar.
  ((1 << c) & invalidChar) === 0
) {
  // do something
}
```

Reply:
I will when I have time.

Reply:
mirco faster

```suggestion
      ((code & -32) === 0 && code !== 0x09)
```

Reply:
@Uzlopak I think the benchmark was wrong and the performance improvement was not detected.
<details>

```js
import { bench, group, run } from "mitata";

const valid = "Space=Cat; Secure; HttpOnly; Max-Age=2";
const invalid = "Space=Cat; Secure; HttpOnly; Max-Age=2\x7F";

function isCTLExcludingHtabCurrent(value) {
  const valueLength = value.length;

  for (let i = 0; i < valueLength; ++i) {
    const code = value.charCodeAt(i);

    if (
      (code >= 0x00 && code <= 0x08) ||
      (code >= 0x0a && code <= 0x1f) ||
      code === 0x7f
    ) {
      return true;
    }
  }
  return false;
}

function isCTLExcludingHtabSuggestion(value) {
  const valueLength = value.length;

  for (let i = 0; i < valueLength; ++i) {
    const code = value.charCodeAt(i);

    if ((code < 0x20 && code !== 0x09) || code === 0x7f) {
      return true;
    }
  }
  return false;
}

function isCTLExcludingHtabNew(value) {
  const valueLength = value.length;
  for (let i = 0; i < valueLength; ++i) {
    const code = value.charCodeAt(i);

    if (((code & -0x20) === 0 && code !== 0x09) || code === 0x7f) {
      return true;
    }
  }
  return false;
}

group(`valid: ${valid}`, () => {
  bench("current", () => isCTLExcludingHtabCurrent(valid));
  bench("suggestion", () => isCTLExcludingHtabSuggestion(valid));
  bench("new", () => isCTLExcludingHtabNew(valid));
});
group(`invalid: ${invalid}`, () => {
  bench("current", () => isCTLExcludingHtabCurrent(invalid));
  bench("suggestion", () => isCTLExcludingHtabSuggestion(invalid));
  bench("new", () => isCTLExcludingHtabNew(invalid));
});

await run();

```

```
benchmark       time (avg)             (min … max)       p75       p99      p999
-------------------------------------------------- -----------------------------
• valid: Space=Cat; Secure; HttpOnly; Max-Age=2
-------------------------------------------------- -----------------------------
current        136 ns/iter       (117 ns … 661 ns)    133 ns    275 ns    480 ns
suggestion  95'864 ps/iter    (83'838 ps … 593 ns) 89'453 ps    225 ns    350 ns
new         93'846 ps/iter    (81'299 ps … 469 ns) 89'502 ps    202 ns    281 ns

summary for valid: Space=Cat; Secure; HttpOnly; Max-Age=2
  new
   1.02x faster than suggestion
   1.45x faster than current

• invalid: Space=Cat; Secure; HttpOnly; Max-Age=2
-------------------------------------------------- -----------------------------
current        136 ns/iter       (113 ns … 466 ns)    138 ns    263 ns    414 ns
suggestion  98'614 ps/iter    (85'742 ps … 372 ns) 91'455 ps    236 ns    323 ns
new         96'604 ps/iter    (83'105 ps … 298 ns) 91'455 ps    193 ns    263 ns

summary for invalid: Space=Cat; Secure; HttpOnly; Max-Age=2
  new
   1.02x faster than suggestion
   1.41x faster than current
```

</details>

Reply:
Sorry. I meant your suggestion.